deepchem/models/torch_models/cnn.py

```diff
@@ -0,0 +1,231 @@
+    self.mode = mode
+    self.layer_filters = layer_filters
+    self.residual = residual
+
+    n_layers = len(layer_filters)
+
+    # PyTorch layers require input and output channels as parameter
+    # if only one layer to make the model creating loop below work, multiply layer_filters wutg 2
+    if len(layer_filters) == 1:
+      layer_filters = layer_filters * 2
+
+    if not isinstance(kernel_size, list):
+      kernel_size = [kernel_size] * n_layers
+    if not isinstance(strides, SequenceCollection):
+      strides = [strides] * n_layers
+    if not isinstance(dropouts, SequenceCollection):
+      dropouts = [dropouts] * n_layers
+    if not isinstance(activation_fns, SequenceCollection):
+      activation_fns = [activation_fns] * n_layers
+
+    if uncertainty:
+
+      if mode != 'regression':
+        raise ValueError("Uncertainty is only supported in regression mode")
+
+      if any(d == 0.0 for d in dropouts):
+        raise ValueError(
+            'Dropout must be included in every layer to predict uncertainty')
+
+    # Python tuples use 0 based indexing, dims defines number of dimension for convolutional operation
+    ConvLayer = (nn.Conv1d, nn.Conv2d, nn.Conv3d)[self.dims - 1]
+
+    PoolLayer: Type
+    if pool_type == 'average':
+      PoolLayer = (nn.AvgPool1d, nn.AvgPool2d, nn.AvgPool3d)[self.dims - 1]
+    elif pool_type == 'max':
+      PoolLayer = (nn.MaxPool1d, nn.MaxPool2d, nn.MaxPool3d)[self.dims - 1]
+    else:
+      raise ValueError("pool_type must be either 'average' or 'max'")
+
+    self.layers = nn.ModuleList()
+
+    in_shape = n_features
+
+    for out_shape, size, stride, dropout, activation_fn in zip(
+        layer_filters, kernel_size, strides, dropouts, activation_fns):
+
+      self.layers.append(
+          ConvLayer(in_channels=in_shape,
+                    out_channels=out_shape,
+                    kernel_size=size,
+                    stride=stride,
+                    padding=padding,
+                    dilation=1,
+                    groups=1,
+                    bias=True))
+
+      if dropout > 0.0:
+        self.layers.append(nn.Dropout(dropout))
+
+      if activation_fn is not None:
+        self.layers.append(activation_fn())
+
+      self.layers.append(PoolLayer(size))
+      in_shape = out_shape
+
+  def forward(self, x: torch.Tensor) -> List[Any]:
+    """
+        Parameters
+        ----------
+        x: torch.Tensor
+          Input Tensor
+        Returns
+        -------
+        torch.Tensor
+          The tensor to be fed into ODEBlock,
+          the length of this tensor is equal to ODEBlock Input Dimension
+        """
+    prev_layer = x
+
+    for layer in self.layers:
+      x = layer(x)
+      # residual blocks can only be used when successive layers have the same output shape
+      if self.residual and layer.in_channels == layer.out_channels:
+        x = x + prev_layer
+      prev_layer = x
+
+    outputs = []
+    batch_size = x.shape[0]
+
```

Comment:
Can you add a comment explaining the use of einops and what the rearrangement here does?